primitives/npos-elections/src/phragmms.rs

```diff
@@ -0,0 +1,381 @@
+	for c_ptr in candidates.iter() {
+		let mut candidate = c_ptr.borrow_mut();
+		if !candidate.elected {
+			candidate.score = Rational128::from(1, P::ACCURACY.into());
+		}
+	}
+
+	for voter in voters.iter() {
+		let mut denominator_contribution: ExtendedBalance = 0;
+
+		// gather contribution from all elected edges.
+		for edge in voter.edges.iter() {
+			let edge_candidate = edge.candidate.borrow();
+			if edge_candidate.elected {
+				let edge_contribution: ExtendedBalance = P::from_rational_approximation(
+					edge.weight,
+					edge_candidate.backed_stake,
+				).deconstruct().into();
+				denominator_contribution += edge_contribution;
+			}
+		}
+
+		// distribute to all _unelected_ edges.
+		for edge in voter.edges.iter() {
+			let mut edge_candidate = edge.candidate.borrow_mut();
+			if !edge_candidate.elected {
+				let prev_d = edge_candidate.score.d();
+				edge_candidate.score = Rational128::from(1, denominator_contribution + prev_d);
+			}
+		}
+	}
+
+	// finalise the score value, and find the best.
+	let mut best_score = Rational128::zero();
+	let mut best_candidate = Rc::clone(&candidates[0]);
+	for c_ptr in candidates.iter() {
+		let mut candidate = c_ptr.borrow_mut();
+		if candidate.approval_stake > 0  {
+			// finalise the score value.
+			let score_d = candidate.score.d();
+			let one: ExtendedBalance = P::ACCURACY.into();
+			let score_n = candidate.approval_stake.checked_mul(one).unwrap_or_else(|| Bounded::max_value());
+			candidate.score = Rational128::from(score_n, score_d);
+
+			// check if we have a new winner.
+			if !candidate.elected && candidate.score > best_score {
+				best_score = candidate.score;
+				best_candidate = Rc::clone(&c_ptr);
+			}
+		} else {
+			candidate.score = Rational128::zero();
+		}
+	}
+
+	best_candidate
+}
+
+/// Update the weights of `voters` given that `elected_ptr` has been elected in the previous round.
+///
+/// Updates `voters` in place.
+///
+/// This is an internal part of the [`balanced_heuristic`] and should be called after
+/// [`calculate_max_score`].
+pub(crate) fn apply_elected<AccountId: IdentifierT>(
+	voters: &mut Vec<Voter<AccountId>>,
+	elected_ptr: CandidatePtr<AccountId>,
+) {
+	let mut elected = elected_ptr.borrow_mut();
+	dbg!(elected.score);
+	let cutoff = elected.score.to_den(1)
+		.expect("(n / d) < u128::max() and (n' / 1) == (n / d), thus n' < u128::max()'; qed.")
+		.n();
+
+	for voter_index in 0..voters.len() {
+		let voter = &mut voters[voter_index];
+
+		for new_edge_index in 0..voter.edges.len() {
+			let new_edge_immutable = &voter.edges[new_edge_index];
+
+
+			// ideally, we'd have to do:
+			// new_edge_immutable.candidate.borrow().who == elected.who
+			// but this will fail because if equality is correct, then borrowing will panic. Hence,
+			// we play with fire here and assume that if `try_borrow()` fails, then it was the same
+			// candidate.
+			if new_edge_immutable.candidate.try_borrow().is_err() {
```

Comment:
I had to do some interesting stuff here that could use a review to prevent double borrow errors.

Reply:
it seems correct to me but rather difficult to read, I also feel some value can be updated at the end, thus we don't need to borrow mutably, I tried a bit it look like this:
```diff
diff --git a/primitives/npos-elections/src/phragmms.rs b/primitives/npos-elections/src/phragmms.rs
index b3bb30e0d..9c9682f92 100644
--- a/primitives/npos-elections/src/phragmms.rs
+++ b/primitives/npos-elections/src/phragmms.rs
@@ -143,66 +143,48 @@ pub(crate) fn apply_elected<AccountId: IdentifierT>(
        voters: &mut Vec<Voter<AccountId>>,
        elected_ptr: CandidatePtr<AccountId>,
 ) {
-       let mut elected = elected_ptr.borrow_mut();
-       let cutoff = elected.score.to_den(1)
+       let elected_who = elected_ptr.borrow().who.clone();
+       let cutoff = elected_ptr.borrow().score.to_den(1)
                .expect("(n / d) < u128::max() and (n' / 1) == (n / d), thus n' < u128::max()'; qed.")
                .n();
 
-       for voter_index in 0..voters.len() {
-               let voter = &mut voters[voter_index];
-
-               for new_edge_index in 0..voter.edges.len() {
-                       let new_edge_immutable = &voter.edges[new_edge_index];
-
-
-                       // ideally, we'd have to do:
-                       // new_edge_immutable.candidate.borrow().who == elected.who
-                       // but this will fail because if equality is correct, then borrowing will panic. Hence,
-                       // we play with fire here and assume that if `try_borrow()` fails, then it was the same
-                       // candidate.
-                       if new_edge_immutable.candidate.try_borrow().is_err() {
-                               let used_budget: ExtendedBalance = voter.edges.iter().map(|e| e.weight).sum();
-                               let new_edge_weight = voter.budget.saturating_sub(used_budget);
-
-                               // just for brevity. We won't need this ref anymore and want to create a new mutable
-                               // one now.
-                               drop(new_edge_immutable);
-
-                               let mut new_edge_mut = &mut voter.edges[new_edge_index];
-                               new_edge_mut.weight = new_edge_weight;
-                               elected.backed_stake = elected.backed_stake.saturating_add(new_edge_weight);
-                               drop(new_edge_mut);
-
-                               for edge_index in 0..voter.edges.len() {
-                                       let mut edge = &mut voter.edges[edge_index];
-
-                                       // opposite of the above error. We need to ensure
-                                       // `edge.candidate.borrow().who != elected.who`, thus if borrowing is okay they
-                                       // are different.
-                                       if edge.weight > 0 && edge.candidate.try_borrow().is_ok() {
-                                               let mut edge_candidate = edge.candidate.borrow_mut();
-
-                                               if edge_candidate.backed_stake > cutoff {
-                                                       let stake_to_take =
-                                                               edge.weight.saturating_mul(cutoff) / edge_candidate.backed_stake.max(1);
-                                                       edge.weight = edge.weight.saturating_sub(stake_to_take);
-                                                       edge_candidate.backed_stake =
-                                                               edge_candidate.backed_stake.saturating_sub(stake_to_take);
-                                                       elected.backed_stake = elected.backed_stake.saturating_add(stake_to_take);
-
-                                                       // drop previous borrows of `voter.edges`.
-                                                       drop(edge_candidate);
-                                                       drop(edge);
-
-                                                       // make another temporary borrow of `voter.edges`.
-                                                       voter.edges[new_edge_index].weight =
-                                                               voter.edges[new_edge_index].weight.saturating_add(stake_to_take);
-                                               }
-                                       }
+       // The backed_stake of elected is updated at the end only.
+       let mut elected_final_backed_stake = elected_ptr.borrow().backed_stake;
+
+       for voter in voters {
+               // Find the index of the edge with new elected candidate
+               if let Some(new_edge_index) = voter.edges.iter().position(|e| e.who == elected_who) {
+                       let used_budget: ExtendedBalance = voter.edges.iter().map(|e| e.weight).sum();
+
+                       // The weight of the edge is updated at after processing other edges.
+                       let mut new_edge_final_weight = voter.budget.saturating_sub(used_budget);
+
+                       // Iter on all other edges
+                       for (_, edge) in voter.edges.iter_mut().enumerate()
+                               .filter(|(edge_index, _)| *edge_index != new_edge_index)
+                       {
+                               if edge.weight == 0 { continue }
+
+                               let mut edge_candidate = edge.candidate.borrow_mut();
+
+                               if edge_candidate.backed_stake > cutoff {
+                                       let stake_to_take =
+                                               edge.weight.saturating_mul(cutoff) / edge_candidate.backed_stake.max(1);
+
+                                       edge.weight = edge.weight.saturating_sub(stake_to_take);
+                                       edge_candidate.backed_stake =
+                                               edge_candidate.backed_stake.saturating_sub(stake_to_take);
+
+                                       new_edge_final_weight = new_edge_final_weight.saturating_add(stake_to_take);
                                }
                        }
+
+                       voter.edges[new_edge_index].weight = new_edge_final_weight;
+                       elected_final_backed_stake =
+                               elected_final_backed_stake.saturating_add(new_edge_final_weight);
                }
        }
+       elected_ptr.borrow_mut().backed_stake = elected_final_backed_stake;
 }
 
 #[cfg(test)]
```

Reply:
Great suggestion. I think it should work too. Albeit when I applied it there were errors and tests were failing, I added your version as comment so you can have a look.

Reply:
Solved, there was one subtle issue with the suggestion otherwise worked fine.